### PR TITLE
Fork vs Spawn in Python Multiprocessing

### PR DESCRIPTION
**Pull request by volcan01010**
Thank you for submitting a pull request for your new post. Please remember to assign the PR for review.

### Blog Post Title

* Fork vs Spawn in Python Multiprocessing

This is a blog post making a fairly in-depth comparison between two ways in which Python can handle parallel processing.
I got stuck with this in my own development on the Ash Model Plotting project and found that there wasn't much good documentation around.

Writing this post help consolidate understanding for me and it will now be there to refer to in the future.
The target audience is intermediate/advanced Python developers who would like to understand better how multiprocessing works.
We can refer to it both internally and externally.

@dvalters has checked an earlier draft of this and may have comments on this version.

### Link to rendered version
Please include a link to the rendered version of your blog post so it can be seen its intended display format,

* https://volcan01010.github.io/BritishGeologicalSurvey.github.io/science/python-forking-vs-spawn/

### CODE REVIEW

**Comment by dvalters** [comment]
It looks good at quick read through and the script runs fine on Linux. (I guess a follow up blog post would be how can we get it to work on windows since it has no 'fork' context). I'm just logging off shortly but will finish the review tomorrow

**Comment by volcan01010** [comment]
Thanks@dvalters. I don't think that Windows users would have the same issue that we did because they get `spawn` by default.  I could emphasise that more.

**Comment by dvalters** [comment]
I ran it on Windows and it just hangs using the `spawn` context...am I missing a quick fix?

**Comment by volcan01010** [comment]
Ah, OK.  I haven't tried it on Windows.  I'll take a look.

**Comment by volcan01010** [comment]
When I try it in Windows, (Python 3.6) it crashes for me with the following:

> Traceback (most recent call last):
  File ".\multi_demo.py", line 95, in <module>
    run_multi(context)
  File ".\multi_demo.py", line 29, in run_multi
    with get_context(context).Pool(2, initializer=init) as pool:
  File "C:\Users\jostev\AppData\Local\Continuum\anaconda3\lib\multiprocessing\context.py", line 238, in get_context
    return super().get_context(method)
  File "C:\Users\jostev\AppData\Local\Continuum\anaconda3\lib\multiprocessing\context.py", line 192, in get_context
    raise ValueError('cannot find context for %r' % method)
ValueError: cannot find context for 'fork'
INFO:multi_demo:*** Lock released in thread process ***

This makes sense as `fork` isn't available on Windows.  What do you get?

**Comment by volcan01010** [comment]
Running with `spawn` worked, but it took a loooooooong time to kick off the new processes.

**Comment by volcan01010** [comment]
I've pushed a new commit to stop the script running in Windows.

**Comment by dvalters** [comment]
Ah, ok - it's a strange Windows IDE bug/behaviour/quirk, I think. Something to do with them not being thread safe but I didn't delve into it very much. (Spyder). The spawn context works fine on the command line. (Yeah it should fall over on `fork` context because it isn't available on Windows. (Works in Linux Spyder though 🤷 ...)

File: _posts/2020-11-13-python-forking-vs-spawn.md [hunk 1]
```diff
@@ -169,9 +169,14 @@ import datetime as dt
...
 from multiprocessing import get_context
 import os
+import sys
 import time
 from threading import Lock, Thread, enumerate
 
+if sys.platform in ('win32', 'msys', 'cygwin'):
+    print("Script works only on *NIX type operating systems.")
+    sys.exit(1)
+
```
Comment (dvalters) [inline _posts/2020-11-13-python-forking-vs-spawn.md:179]: I suppose if we wanted some windows compatibility, we could have a context based on the platform eg. ['spawn'] for windows and ['spawn', 'fork'] for nix. (Although maybe that's making it too complicated and missing the point of the post.)

**Review by dvalters — APPROVED** [review]
It looks good!

File: _posts/2020-11-13-python-forking-vs-spawn.md (hunk 1 above)
Comment (volcan01010) [reply _posts/2020-11-13-python-forking-vs-spawn.md:179]: I had weirdness when I ran it in Spyder in Windows, too.  🤷

The main aim is to show the difference between fork and spawn and explain why the hanging thread issue happens.  Windows doesn't have fork or the hanging thread problem, so I think it's OK just to stop the script at the start for them.

Comment (dvalters) [reply _posts/2020-11-13-python-forking-vs-spawn.md:179]: yes, that seems reasonable 👍

**Comment by pdbebgs** [comment]
Signed off by Patrick

**Pull request merged**